File: Mushroom/Cells/Water.cs
using System;
using Mushroom.Data;
using Godot;

namespace Mushroom.Ceils;

public class Water : CellBase
{
    private int lifeTime = 0;

    public override Action? Do(Vector2I position)
    {
        int nextLifeTime = lifeTime;
        
        Direction[] directions = { Direction.Up, Direction.Down, Direction.Left, Direction.Right };
        int waterCeils = 0;
        bool isAir = false;
        
        foreach (var dir in directions)
        {
            var neighbor = Grid.GetNeighbor(position, dir);
            if (neighbor is Water) 
                waterCeils++;
            if (neighbor is Air) 
                isAir = true;

            if (neighbor is Dirt dirt && dirt.Dampness < 1f)
            {
                return () =>
                {
                    dirt.Dampness = Math.Clamp(dirt.Dampness + 0.3f, 0f, 1f);
                    Grid.Set(position, Air.Instance);
                };
            }
        }

        if (isAir && waterCeils < 3) 
            nextLifeTime++;

        if (nextLifeTime > 300)
            return () => Grid.Set(position, Air.Instance);

        var downPosition = new Vector2I(position.X, position.Y + 1);
        if (Grid.Get(downPosition) is Air)
            return () =>
            {
                lifeTime = nextLifeTime; 
                Grid.Move(position, downPosition);
            };
        
        var downLeftPosition = new Vector2I(position.X - 1, position.Y + 1);
        var downRightPosition = new Vector2I(position.X + 1, position.Y + 1);

        bool canFlowLeftDown = Grid.Get(downLeftPosition) is Air;
        bool canFlowRightDown = Grid.Get(downRightPosition) is Air;
        
        if (canFlowLeftDown && canFlowRightDown)
        {
            var targetPosition = Random.Shared.Next(0, 2) == 0 ? downLeftPosition : downRightPosition;
            return () =>
            {
                lifeTime = nextLifeTime; 
                Grid.Move(position, targetPosition);
            };
        }
        if (canFlowLeftDown) 
            return () =>
            {
                lifeTime = nextLifeTime; 
                Grid.Move(position, downLeftPosition);
            };
        if (canFlowRightDown) 
            return () => 
            { 
                lifeTime = nextLifeTime; 
                Grid.Move(position, downRightPosition);
            };
        
        var leftPosition = new Vector2I(position.X - 1, position.Y);
        var rightPosition = new Vector2I(position.X + 1, position.Y);

        bool canMoveLeft = Grid.Get(leftPosition) is Air;
        bool canMoveRight = Grid.Get(rightPosition) is Air;
        
        if (canMoveLeft && canMoveRight)
        {
            var targetPosition = Random.Shared.Next(0, 2) == 0 ? leftPosition : rightPosition;
            return () =>
            {
                lifeTime = nextLifeTime; 
                Grid.Move(position, targetPosition);
            };
        }

        if (canMoveLeft) 
            return () =>
            {
                lifeTime = nextLifeTime; 
                Grid.Move(position, leftPosition);
            };
        if (canMoveRight) 
            return () =>
            {
                lifeTime = nextLifeTime; 
                Grid.Move(position, rightPosition);
            };
        

        if (nextLifeTime != lifeTime)
            return () =>
            {
                lifeTime = nextLifeTime;
            };

        return null;
    }

    public override Color GetColor(Vector2I position)
    {
        //int depth = 0;
        //var currentPos = new Vector2I(position.X, position.Y + 1);

        // while (Grid.Get(currentPos) is Water && depth < 10)
        // {
        //     depth++;
        //     currentPos = new Vector2I(currentPos.X, currentPos.Y + 1);
        // }

        Color baseColor = new Color(0.33f, 0.66f, 0.9f);

        //baseColor.R = Math.Max(0, baseColor.R - depth * 15);
        //baseColor.G = Math.Max(0, baseColor.G - depth * 15);
        //baseColor.B = Math.Max(0, baseColor.B - depth * 10);

        return baseColor;
    }

    public override Color GetUiColor()
        => new Color(0.33f, 0.66f, 0.9f);
}

File: Core/src/cells/water.rs
use easy_color::{Hex, IntoRGB, RGB};
use crate::cells::dirt::Dirt;
use crate::cells::air::Air;
use strum::IntoEnumIterator;
use crate::cells;
use crate::cells::{Action, Cell, AsAny};
use crate::data::{Direction, Position};
use crate::grid::grid;
use std::sync::{Arc, Mutex};
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};

#[derive(Clone)]
pub struct Water {
    pub life_time: Arc<Mutex<i32>>,
    rnd: Arc<Mutex<StdRng>>
}

impl Water {
    pub fn new() -> Self {
        Self {
            life_time: Arc::new(Mutex::new(0)),
            rnd: Arc::new(Mutex::new(StdRng::from_os_rng()))
        }
    }
}

impl Cell for Water {
    fn do_action(&self, position: Position) -> Option<Action> {
        let mut waterCells = 0;
        let mut isAir = false;

        for dir in Direction::iter() {

            if let Some(neighbor) = grid::get_neighbor(position, dir) {
                if neighbor.as_any().is::<Water>() {
                    waterCells += 1;
                }
                else if neighbor.as_any().is::<Air>() {
                    isAir = true;
                }
                else if neighbor.as_any().is::<Dirt>() {
                    let cloned_neighbor = neighbor.clone();
                    if let Some(target_dirt) = neighbor.as_any().downcast_ref::<Dirt>() {
                        let target_dampness = target_dirt.dampness.lock().unwrap();
                        if target_dampness.clone() < 1.0 {
                            return Some(Box::new(move || {
                                if let Some(dirt) = cloned_neighbor.as_any().downcast_ref::<Dirt>() {
                                    let mut dampness = dirt.dampness.lock().unwrap();

                                    let value = dampness.clone();
                                    *dampness = (value + 0.3);
                                }
                            }));
                        }
                    }
                }
            }
        }

        let mut life_time = self.life_time.lock().unwrap();
        if isAir && waterCells < 3 {
            *life_time = *life_time + 1;
        }

        if *life_time > 300 {
            return Some(Box::new(move || {
                grid::set(position, Arc::new(Air { }))
            }))
        }

        let down_position = Position::new(position.x, position.y + 1);
        if grid::get(down_position).as_any().is::<Air>() {
            return Some(Box::new(move || {
                grid::move_cell(position, down_position)
            }))
        }

        let down_left_position = Position::new(position.x - 1, position.y + 1);
        let down_right_position = Position::new(position.x + 1, position.y + 1);

        let can_flow_left_down = grid::get(down_left_position).as_any().is::<Air>();
        let can_flow_right_down = grid::get(down_right_position).as_any().is::<Air>();

        let mut rand = self.rnd.lock().unwrap();
        if can_flow_left_down && can_flow_right_down {
            let target_position = if rand.random_range(0..=1) == 0 { down_left_position } else { down_right_position };
            return Some(Box::new(move || {
                grid::move_cell(position, target_position)
            }))
        }
        if can_flow_left_down {
            return Some(Box::new(move || {
                grid::move_cell(position, down_left_position)
            }))
        }
        if can_flow_right_down {
            return Some(Box::new(move || {
                grid::move_cell(position, down_right_position)
            }))
        }

        Some(Box::new(move || {}))
    }

    fn get_color(&self, _position: Position) -> String {
        let mut depth = 0;
        let mut currentPos = Position::new(_position.x, _position.y + 1);

        while grid::get(currentPos).as_any().is::<Water>() && depth < 10 {
            depth += 1;
            currentPos = Position::new(currentPos.x, currentPos.y + 1);
        }
        let baseColor: Hex = "#55a8e8".try_into().unwrap();
        let mut rgb: RGB = baseColor.into();

        rgb.set_red((rgb.red() - depth * 15).max(0));
        rgb.set_green((rgb.green() - depth * 15).max(0));
        rgb.set_blue((rgb.blue() - depth * 10).max(0));

        Hex::from(rgb).to_string()
    }
    fn clone_cell(&self) -> Arc<(dyn cells::Cell + 'static)> {
        let new_water = Water {
            life_time: Arc::new(Mutex::new(*self.life_time.lock().unwrap())),
            rnd: self.rnd.clone()
        };
        Arc::new(new_water)
    }
}